rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,116 @@
+  namespace: teleport
+spec:
+  allow:
+    request:
+      roles: ["admin"]
+---
+apiVersion: resources.goteleport.com/v4
+kind: User
+metadata:
+  name: human-bean
+  namespace: teleport
+spec:
+  roles: ["folks"]
+```
+
+Manager process watches for custom resource events. Once a custom resource is created/updated, the manager's watcher acknowledges it and upserts a resource into Teleport.
+
+The name of the Kubernetes custom resource is the name of the resource in Teleport. Also, every Kubernetes custom resource must have a `spec` field that has the same schema as a `spec` field of a resource in Teleport.
+
+Deletion of resources is being handled using [resource finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). Every acknowledged resource gets a finalizer named `"resources.goteleport.com/delete"` so when the deletion happens, Kubernetes doesn't actually delete it but sets a deletion timestamp. While the finalizer list is non-empty, the manager can still access the resource's fields. Deletion of the Kubernetes object leads to deletion of the associated resource in Teleport. Once a deletion in Teleport succeeds, the manager removes a finalizer from the Kubernetes object so it can be garbage-collected.
+
+### Identity definitions
+
+To run some plugin in the cluster, the credentials are required. Normally, one can get the credentials using `tctl auth sign` command or using `tsh login`. To automate the credentials generation, the `Identity` resource is introduced. The identity definition looks like this:
+
+```yaml
+apiVersion: control.goteleport.com/v8
+kind: Identity
+metadata:
+  name: access-slack
+spec:
+  username: access-slack
+  secretName: access-slack-secret
+```
+
+Here, the `username` is a Teleport user name, and `secretName` is a name of a secret resource where the manager will write credentials. Once the `Identity` object is created in the Kubernetes, the secret object with credentials contents will be written in the same namespace where the identity object resides.
+
+Once an identity is removed, any secret data it had generated should remain untouched. However, if the user wants the identity secret to being deleted as well, there should be a configuration option for this:
+
+
+```yaml
+identities:
+  ownerRefEnabled: true
+```
+
+When `ownerRefEnabled` is true, the manager sets a [metadata.ownerReferences](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/) of a `Secret` object referenced by `secretName` of some `Identity`. Once an `Identity` resource is removed, Kubernetes will automatically clean up the referenced secret object.
+
+By default, `ownerRefEnabled` should be turned off because if the user suddenly deleted the `Identity` object it can break the deployment.
+
+If the secret referenced by `secretName` has been removed or its secret data has been overwritten, it's the manager's responsibility to re-generate the identity and to write it back.
```

Comment:
> it's the manager's responsibility to re-generate the identity and to write it back.

Who's the "manager" in this context? Human operator or this K8s operator? Can you clarify this in the doc.

Reply:
The "manager" word is used everywhere in the text meaning the k8s controller. The RFD itself is called "Kubernetes manager" and the definition of it was given in a "What" section.

Reply:
The word "operator" is confusing too. The word "controller" is good but typically managers consist of multiple controllers. You've got `kube-controller-manager`, `cert-manager` etc in your cluster.